components/toast_wrapper/toast_wrapper.tsx

```diff
@@ -2,103 +2,120 @@
-        scrollToNewMessage: PropTypes.func,
-        scrollToLatestMessages: PropTypes.func,
-        scrollToUnreadMessages: PropTypes.func,
-        updateLastViewedBottomAt: PropTypes.func,
-        showSearchHintToast: PropTypes.bool,
-        onSearchHintDismiss: PropTypes.func,
-        shouldStartFromBottomWhenUnread: PropTypes.bool,
-        isNewMessageLineReached: PropTypes.bool,
-        unreadScrollPosition: PropTypes.string,
-
-        /*
-         * Object from react-router
-         */
-        match: PropTypes.shape({
-            params: PropTypes.shape({
-                team: PropTypes.string,
-            }).isRequired,
-        }).isRequired,
-
-        actions: PropTypes.shape({
-
-            /**
-             * Action creator to update toast status
-             */
-            updateToastStatus: PropTypes.func.isRequired,
-        }).isRequired,
+type OwnProps = {
+    unreadCountInChannel: number;
+    newRecentMessagesCount: number;
+    channelMarkedAsUnread?: boolean;
+    isCollapsedThreadsEnabled?: boolean;
+    rootPosts?: IDMappedObjects<Post>;
+    atLatestPost?: boolean;
+    postListIds: string[];
+    latestPostTimeStamp?: number;
+    atBottom?: boolean | null;
+    lastViewedBottom: number;
+    width: number;
+    lastViewedAt: number;
+    focusedPostId?: string;
+    initScrollOffsetFromBottom?: number;
+    updateNewMessagesAtInChannel: (lastViewedAt?: number) => void;
+    scrollToNewMessage: () => void;
+    scrollToLatestMessages: () => void;
+    scrollToUnreadMessages?: () => void;
+    updateLastViewedBottomAt: (lastViewedBottom?: number) => void;
+    showSearchHintToast?: boolean;
+    onSearchHintDismiss?: () => void;
+    shouldStartFromBottomWhenUnread?: boolean;
+    isNewMessageLineReached?: boolean;
+    unreadScrollPosition?: string;
+    intl: IntlShape;
+    match: {
+        params: {
+            team?: string;
+        };
     };
-
+    actions: {
+        updateToastStatus: (toastPresent: boolean) => void;
+    };
+};
+
+type State = {
+    unreadCount?: number;
+    unreadCountInChannel: number;
+    channelMarkedAsUnread?: boolean;
+    lastViewedAt?: number;
+    showUnreadToast?: boolean;
+    showNewMessagesToast?: boolean;
+    showMessageHistoryToast?: boolean;
+    showUnreadWithBottomStartToast?: boolean;
+};
+
+export type Props = OwnProps;
+
+export class ToastWrapper extends React.PureComponent<Props, State> {
+    mounted: boolean;
     static defaultProps = {
         focusedPostId: '',
     };
 
-    constructor(props) {
+    constructor(props: Props) {
         super(props);
+        this.mounted = true;
         this.state = {
             unreadCountInChannel: props.unreadCountInChannel,
         };
     }
 
-    static countNewMessages = (postListIds, rootPosts, isCollapsedThreadsEnabled) => {
+    static countNewMessages = (postListIds: string[], rootPosts: IDMappedObjects<Post>, isCollapsedThreadsEnabled: boolean) => {
```

Comment:
This should eliminate the need for castings on line 118. 
```suggestion
    static countNewMessages = (postListIds: string[], rootPosts: Record<string, boolean>, isCollapsedThreadsEnabled?: boolean) => {
```